### PR TITLE
[fix](daemon) Fix incorrect initialization order of daemon services

### PR DESCRIPTION
**Pull request by platoneko**
## Proposed changes

Current initialization dependency:

```
      Daemon ───┬──► StorageEngine ──► ExecEnv ──► Disk/Mem/CpuInfo
                │
                │
BackendService ─┘
```
However, original code incorrectly initialize Daemon before StorageEngine.
This PR also stop and join threads of daemon services in their dtor, to ensure Daemon services release resources in reverse order of initialization via RAII.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by platoneko** [comment]
run build_all

**Comment by platoneko** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.63 seconds
 stream load tsv:          548 seconds loaded 74807831229 Bytes, about 130 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          28.6 seconds inserted 10000000 Rows, about 349K ops/s
 storage size: 17162008677 Bytes

File: be/src/runtime/memory/thread_mem_tracker_mgr.h [hunk 1]
```diff
@@ -164,7 +162,7 @@ inline void ThreadMemTrackerMgr::pop_consumer_tracker() {
 
 inline void ThreadMemTrackerMgr::consume(int64_t size, bool large_memory_check) {
     _untracked_mem += size;
-    if (!k_doris_run || !ExecEnv::GetInstance()->initialized()) {
```
Comment (yiguolei) [inline be/src/runtime/memory/thread_mem_tracker_mgr.h:167]: @xinyiZzz  could we remove this variable???

File: be/src/service/http_service.cpp [hunk 2]
```diff
@@ -63,7 +63,9 @@ HttpService::HttpService(ExecEnv* env, int port, int num_threads)
           _ev_http_server(new EvHttpServer(port, num_threads)),
           _web_page_handler(new WebPageHandler(_ev_http_server.get())) {}
 
-HttpService::~HttpService() {}
+HttpService::~HttpService() {
+    stop();
```
Comment (yiguolei) [inline be/src/service/http_service.cpp:67]: if the stop method is called inside this class. then it should be a prviate method.

**Comment by platoneko** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.44 seconds
 stream load tsv:          549 seconds loaded 74807831229 Bytes, about 129 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          28.7 seconds inserted 10000000 Rows, about 348K ops/s
 storage size: 17162044698 Bytes

**Review by yiguolei — DISMISSED** [review]
LGTM

File: be/src/runtime/memory/thread_mem_tracker_mgr.h (hunk 1 above)
Comment (xinyiZzz) [reply be/src/runtime/memory/thread_mem_tracker_mgr.h:167]: It's OK, `ExecEnv.ready` is a static variable
Previously `ExecEnv::GetInstance()` would create `ExecEnv` object

File: be/src/runtime/thread_context.h [hunk 3]
```diff
@@ -238,7 +235,7 @@ class SwitchBthreadLocal {
                 // The brpc server should respond as quickly as possible.
                 bthread_context->thread_mem_tracker_mgr->disable_wait_gc();
                 // set the data so that next time bthread_getspecific in the thread returns the data.
-                CHECK((0 == bthread_setspecific(btls_key, bthread_context)) || doris::k_doris_exit);
+                CHECK((0 == bthread_setspecific(btls_key, bthread_context)));
```
[the remark this answers is not in the document]
Comment (platoneko) [reply be/src/runtime/thread_context.h:238]: Understand

**Comment by platoneko** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 48.49 seconds
 stream load tsv:          538 seconds loaded 74807831229 Bytes, about 132 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17162038646 Bytes

**Comment by yiguolei** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.73 seconds
 stream load tsv:          554 seconds loaded 74807831229 Bytes, about 128 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          28.9 seconds inserted 10000000 Rows, about 346K ops/s
 storage size: 17161887684 Bytes

**Review by yiguolei — APPROVED** [review]
LGTM

**Review by ByteYue — APPROVED** [review]

**Comment by yiguolei** [comment]
run clickbench

**Comment by yiguolei** [comment]
run ckbench

**Pull request merged**